bson/bsoncodec/byte_slice_codec.go

```diff
@@ -40,11 +44,11 @@ func NewByteSliceCodec(opts ...*bsonoptions.ByteSliceCodecOptions) *ByteSliceCod
 }
 
 // EncodeValue is the ValueEncoder for []byte.
-func (bsc *ByteSliceCodec) EncodeValue(_ EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
+func (bsc *ByteSliceCodec) EncodeValue(ec EncodeContext, vw bsonrw.ValueWriter, val reflect.Value) error {
 	if !val.IsValid() || val.Type() != tByteSlice {
 		return ValueEncoderError{Name: "ByteSliceEncodeValue", Types: []reflect.Type{tByteSlice}, Received: val}
 	}
-	if val.IsNil() && !bsc.EncodeNilAsEmpty {
+	if val.IsNil() && !bsc.EncodeNilAsEmpty && !ec.NilByteSliceAsEmpty {
```

Comment:
This test appears to fail for me:

```go
// Marshal with EncodeContext
{
	ectx := bsoncodec.EncodeContext{Registry: bson.DefaultRegistry, NilByteSliceAsEmpty: true}
	got, err := bson.MarshalWithContext(ectx, f)
	if err != nil {
		log.Fatalf("error in Marshal: %v", err)
	}
	var decoded bson.M
	bson.Unmarshal(got, &decoded)

	if _, ok := decoded["mybytes"].(primitive.Binary); !ok {
		log.Fatalf("expected primitive.Binary, got: %T", decoded["mybytes"]) // Results in 'expected primitive.Binary, got: <nil>'
	}
}
```

Is the test incorrect? Does `NilByteSliceAsEmpty` need to be applied elsewhere?